File: ios/StarXpandCommand/BuzzerBuilderWrapper.m
#import "BuzzerBuilderWrapper.h"
#import "StarObjectManager.h"
#import "StarIO10ValueConverter.h"
#import <React/RCTLog.h>

@import StarIO10;

@interface BuzzerBuilderWrapper()

@property(assign, nonatomic) StarObjectManager *objManager;

@end


@implementation BuzzerBuilderWrapper

- (instancetype)init
{
    self = [super init];
    if (self) {
        _objManager = StarObjectManager.sharedManager;
    }
    return self;
}

+ (BOOL)requiresMainQueueSetup
{
    return NO;
}

RCT_EXPORT_MODULE()

RCT_REMAP_METHOD(init,
                 createWithResolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
    STARIO10StarXpandCommandBuzzerBuilder *builder = [[STARIO10StarXpandCommandBuzzerBuilder alloc] init];
    
    if (builder == nil) {
        reject(@"Error", @"Fail to create object.", nil);
        return;
    }
    
    NSString *objID = [_objManager add:builder];
    
    resolve(objID);
}

RCT_REMAP_METHOD(dispose,
                 disposeWithObjectIdentifier:(nonnull NSString *)objID
                 resolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
    [_objManager remove:objID];
    
    resolve(nil);
}

RCT_REMAP_METHOD(actionDrive,
                 actionDriveWithObjectIdentifier:(nonnull NSString *)objID
                 channel:(nonnull NSString *)channel
                 repeat:(nonnull NSNumber *)repeat
                 onTime:(nonnull NSNumber *)onTime
                 offTime:(nonnull NSNumber *)offTime
                 resolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
    STARIO10StarXpandCommandBuzzerBuilder *builder = [_objManager getObject:objID];
    
    if (builder == nil) {
        reject(@"Error", @"Fail to get object.", nil);
        return;
    }
    
    STARIO10StarXpandCommandBuzzerDriveParameter *param = [StarIO10ValueConverter toBuzzerDriveParameterWithChannel:channel repeat:repeat onTime:onTime offTime:offTime];
    
    [builder actionDrive:param];
    
    resolve(nil);
}

@end

File: android/src/main/java/com/stario10module/BuzzerBuilderWrapper.kt
package com.stario10module

import com.facebook.react.bridge.*
import com.starmicronics.stario10.starxpandcommand.BuzzerBuilder

class BuzzerBuilderWrapper internal constructor(context: ReactApplicationContext) : ReactContextBaseJavaModule(context) {
    override fun getName(): String {
        return "BuzzerBuilderWrapper"
    }

    @ReactMethod
    fun init(promise: Promise) {
        val builder = BuzzerBuilder()
        val identifier = InstanceManager.set(builder)

        promise.resolve(identifier)
    }

    @ReactMethod
    fun dispose(identifier: String, promise: Promise) {
        InstanceManager.remove(identifier)
        promise.resolve(0)
    }

    @ReactMethod
    fun actionDrive(identifier: String, channel: String, repeat: Int, onTime: Int, offTime: Int, promise: Promise) {
        val builder = InstanceManager.get(identifier)

        if (builder is BuzzerBuilder) {
            val parameter = StarIO10ValueConverter.toBuzzerDriveParameter(channel, repeat, onTime, offTime)

            builder.actionDrive(parameter)
            promise.resolve(0)
        }
        else {
            promise.reject(ReactNoCrashSoftException("Not found native instance"))
        }
    }
}